tests/Unit/AutoRouteManagerTest.php

```diff
@@ -11,154 +11,360 @@
-        $this->uriGenerator->generateUri($this->context1->reveal())->willReturn($uri);
-        $this->context1->setUri($uri)->shouldBeCalled();
-        $this->context1->getLocale()->willReturn(null);
-        $this->context1->getSubject()->willReturn($this->subject);
-        $this->adapter->findRouteForUri($uri, $this->context1)->willReturn(
-            $this->autoRoute1->reveal()
-        );
+        $tag = 'tag';
+        $subject = $context->reveal()->getSubject();
+
+        $this->adapter->generateAutoRouteTag($context)->willReturn($tag);
+
+        $this->adapter->createAutoRoute($context, $subject, $tag)
+            ->willReturn($this->prophesize(AutoRouteInterface::class));
 
-        // handle existing route
-        $this->adapter->compareAutoRouteContent(
-            $this->autoRoute1->reveal(),
-            $this->subject
-        )->willReturn($sameContent);
+        if ($route['existsInDatabase']) {
+            $existingAutoRoute = $this->prophesize(AutoRouteInterface::class);
 
-        $this->context1->getSubject()->willReturn($this->subject);
+            $this->adapter->findRouteForUri(
+                $route['generatedUri'],
+                $context
+            )->willReturn($existingAutoRoute);
 
-        if ($sameContent) {
-            $this->autoRoute1->setType(AutoRouteInterface::TYPE_PRIMARY)
-                ->shouldBeCalled();
+            $this->adapter->compareAutoRouteContent(
+                $existingAutoRoute->reveal(),
+                $subject
+            )->willReturn($route['withSameContent'] === true);
         } else {
-            $this->uriGenerator->resolveConflict($this->context1->reveal())
-                ->willReturn($resolvedUri);
-            $this->context1->setUri($resolvedUri)->shouldBeCalled();
-            $this->adapter->generateAutoRouteTag($this->context1->reveal())->willReturn('fr');
-            $this->adapter->createAutoRoute($this->context1, $this->subject, 'fr')->willReturn($this->autoRoute1->reveal());
+            $this->adapter->findRouteForUri($route['generatedUri'], $context)->willReturn(null);
         }
 
-        $this->context1->setAutoRoute($this->autoRoute1->reveal())->shouldBeCalled();
+        if (!is_null($route['locale'])) {
+            $translatedSubject = new \stdClass();
 
-        $this->manager->buildUriContextCollection($this->collection->reveal());
+            $this->adapter->translateObject($subject, $route['locale'])->willReturn($translatedSubject);
+        } else {
+            $this->adapter->translateObject($subject, $route['locale'])->willReturn($subject);
+        }
     }
 
-    public function provideBuildCollectionExisting()
+    /**
+     * @testdox builds the collection with
+     * @dataProvider routesConfigurations
+     */
+    public function buildUriContextCollection($routes)
     {
-        return [
-            [true],
-            [false],
-        ];
+        $collection = $this->prophesize(UriContextCollection::class);
```

Comment:
Yes, using the real classes would make the test more maintainable. But by using the real classes, we rely on their implementation. I think stub are made to avoid it.